src/python/pants/engine/legacy/graph.py

```diff
@@ -430,6 +578,17 @@ def create_legacy_graph_tasks(symbol_table):
         Get(HydratedField, BundlesField),
       ]
     ),
+    TaskRule(
```

Comment:
When will this `TaskRule` thing go away? It feels like it's here because `find_owners` isn't decorated with `@rule`? Is that because params aren't properly injectable or something?

I don't like how much of this duplicates information from the body of the function itself...

Reply:
> Is that because params aren't properly injectable or something?

Correct. This is here because of `symbol_table_constraint`, which is the same underlying problem as #4535. Need a sealed universe of types that are legal in a particular position: in this case, we manually build up and pass around the set of types that can legally be parsed from a `BUILD` file. I'll add a comment on #4535 to that effect.